pkg/env/certificateRefresher_test.go

```diff
@@ -0,0 +1,283 @@
+			sleep:    60 * time.Millisecond,
+			managerFactory: func(controller *gomock.Controller) keyvault.Manager {
+				manager := mock_keyvault.NewMockManager(mockController)
+				gomock.InOrder(
+					manager.EXPECT().GetCertificateSecret(gomock.Any(), testCertName).Return(key1, certs1, nil),
+					manager.EXPECT().GetCertificateSecret(gomock.Any(), testCertName).Return(key2, certs2, nil),
+				)
+				return manager
+			},
+			wantKey:  key2,
+			wantCert: certs2[0],
+			wantErr:  nil,
+		},
+		{
+			name:     "test initial error, pull exactly once with an error",
+			interval: 1 * time.Minute,
+			sleep:    20 * time.Millisecond,
+			managerFactory: func(controller *gomock.Controller) keyvault.Manager {
+				manager := mock_keyvault.NewMockManager(mockController)
+				manager.EXPECT().GetCertificateSecret(gomock.Any(), testCertName).Return(nil, nil, cannotPull)
+				return manager
+			},
+			wantKey:  nil,
+			wantCert: nil,
+			wantErr:  cannotPull,
+		},
+		{
+			name:     "test refresh error, pull exactly twice, first on start, second time with an error",
+			interval: 50 * time.Millisecond,
+			sleep:    60 * time.Millisecond,
+			managerFactory: func(controller *gomock.Controller) keyvault.Manager {
+				manager := mock_keyvault.NewMockManager(controller)
+				gomock.InOrder(
+					manager.EXPECT().GetCertificateSecret(gomock.Any(), testCertName).Return(key1, certs1, nil),
+					manager.EXPECT().GetCertificateSecret(gomock.Any(), testCertName).Return(nil, nil, cannotPull),
+				)
+				return manager
+			},
+			wantKey:  key1,
+			wantCert: certs1[0],
+			wantErr:  nil,
+		},
+		{
+			name:     "test refresh, pull exactly 10 times",
+			interval: 10 * time.Millisecond,
+			sleep:    100 * time.Millisecond,
+			managerFactory: func(controller *gomock.Controller) keyvault.Manager {
+				manager := mock_keyvault.NewMockManager(controller)
+				manager.EXPECT().GetCertificateSecret(gomock.Any(), testCertName).Return(key1, certs1, nil).Times(10)
+				return manager
+			},
+			wantKey:  key1,
+			wantCert: certs1[0],
+			wantErr:  nil,
+		},
+	}
+
+	for _, test := range tt {
+		t.Run(test.name, func(t *testing.T) {
+			fmt.Println(test.name)
+			stop := make(chan struct{})
+			refreshing := newCertificateRefresher(
+				context.Background(),
+				logrus.NewEntry(logrus.StandardLogger()),
+				test.interval,
+				test.managerFactory(mockController),
+				testCertName,
+				stop,
+			)
+
+			err := refreshing.Start(context.Background())
+			defer close(stop)
+			if err != test.wantErr {
+				t.Fatalf("unexpected error occurred when starting the poller, cannot get initial certificate: %s", err.Error())
+			}
+
+			if err != nil {
+				return
+			}
+
+			time.Sleep(test.sleep)
+
+			testkey, testCerts := refreshing.GetCertificates()
+			if !equalPrivKey(testkey, test.wantKey) {
+				t.Fatal("returned private key does not match")
+			}
+
+			if !testCerts[0].Equal(test.wantCert) {
+				t.Fatal("returned certificate does not match")
+			}
```

Comment:
I think it makes sense to use `t.Error` in checks: while it is unlikely that `GetCertificates` is going to return valid private key, but invalid cert... I think it is still good idea to run all all the checks.



```suggestion
			testkey, testCerts := refreshing.GetCertificates()
			if !equalPrivKey(testkey, test.wantKey) {
				t.Error("returned private key does not match")
			}

			if !testCerts[0].Equal(test.wantCert) {
				t.Error("returned certificate does not match")
			}
```